src/plugins/data/public/query/query_string/dataset_service/lib/index_data_structure_creator/index_data_structure_creator.tsx

```diff
@@ -3,131 +3,253 @@
 import './index_data_structure_creator.scss';
 
 type SelectionMode = 'single' | 'prefix';
 
-export const IndexDataStructureCreator: React.FC<DataStructureCreatorProps> = ({
+interface ResolveIndexResponse {
+  indices?: Array<{ name: string; attributes?: string[] }>;
+  aliases?: Array<{ name: string }>;
+  data_streams?: Array<{ name: string }>;
+}
+
+interface IndexDataStructureCreatorProps extends DataStructureCreatorProps {
+  services?: IDataPluginServices;
+}
+
+export const IndexDataStructureCreator: React.FC<IndexDataStructureCreatorProps> = ({
   path,
   index,
   selectDataStructure,
+  services,
 }) => {
   const current = path[index];
-  const isLast = index === path.length - 1;
-  const isFinal = isLast && !current.hasNext;
 
   const [selectionMode, setSelectionMode] = useState<SelectionMode>('single');
-  const [customPrefix, setCustomPrefix] = useState('');
-  const [validationError, setValidationError] = useState<string>('');
-  const [appendedWildcard, setAppendedWildcard] = useState(false);
-  const [selectedIndexId, setSelectedIndexId] = useState<string | null>(null);
-
-  // Filter indices that match the custom prefix pattern
-  const matchingIndices = useMemo(() => {
-    if (selectionMode !== 'prefix' || !customPrefix) {
-      return [];
+  const [selectedIndexIds, setSelectedIndexIds] = useState<string[]>([]);
+  const [wildcardPatterns, setWildcardPatterns] = useState<string[]>([]);
+  const [currentWildcardPattern, setCurrentWildcardPattern] = useState('');
+  const [matchingIndices, setMatchingIndices] = useState<string[]>([]);
+  const [isLoadingMatches, setIsLoadingMatches] = useState(false);
+  const debounceTimerRef = useRef<NodeJS.Timeout | null>(null);
+
+  // Fetch indices matching a wildcard pattern using API
+  const fetchMatchingIndices = useCallback(
+    async (pattern: string) => {
+      if (!services?.http || !pattern || pattern.trim() === '') {
+        setMatchingIndices([]);
+        return;
+      }
+
+      try {
+        setIsLoadingMatches(true);
+
+        const dataSourceId = path.find((item) => item.type === 'DATA_SOURCE')?.id;
+        const query: any = {
+          expand_wildcards: 'all',
+        };
+
+        if (dataSourceId && dataSourceId !== '') {
+          query.data_source = dataSourceId;
+        }
+
+        // Check if pattern contains commas (multiple patterns)
+        const patterns = pattern
+          .split(',')
+          .map((p) => p.trim())
+          .filter((p) => p);
+
+        // Fetch matches for all patterns in parallel
+        const allResponses = await Promise.all(
+          patterns.map(
+            (p) =>
+              services.http
+                .get<ResolveIndexResponse>(
+                  `/internal/index-pattern-management/resolve_index/${encodeURIComponent(p)}`,
+                  { query }
+                )
+                .catch(() => null) // Handle individual pattern failures gracefully
+          )
+        );
```

Comment:
We should add console.error and toast notification if the fetching fails. 

```
  } catch (error) {
    // On error, set empty array
    setMatchingIndices([]);
  }

Reply:
Refactored this into a shared useIndexFetcher hook that includes comprehensive error handling:

  1. Individual pattern failures: Each failed pattern is caught and logged with console.error for debugging (lines 64-69 in use_index_fetcher.ts)
  2. Complete fetch failures: The catch block (lines 109-134) includes:
    - console.error for debugging
    - Toast notification for user feedback with i18n messages

  The hook also handles race conditions using requestIdRef to ignore stale requests.

  Created comprehensive test coverage with 13 tests in use_index_fetcher.test.ts including specific tests for:
  - Individual pattern failures returning partial results
  - All patterns failing returning empty array
  - Error logging verification